proxy-server/background/proxy-handler.js

```diff
@@ -0,0 +1,114 @@
+// Log any errors from the proxy script
+browser.proxy.onProxyError.addListener(error => {
+  console.error(`Proxy error: ${error.message}`);
+});
+
+function logSettings(storedSettings) {
+  console.log("Got proxy server info and auth info:");
+  console.log(`proxy server address: ${storedSettings.proxyServerInfo.proxyServerAddress}`);
+  console.log(`proxy server port: ${storedSettings.proxyServerInfo.proxyServerPort}`);
+  console.log(`proxy server type: ${storedSettings.proxyServerInfo.proxyServerType}`);
+  console.log(`proxy server username: ${storedSettings.authCredentials.username}`);
+  console.log(`proxy server password: ${storedSettings.authCredentials.password}`);
+}
+
+let proxyScriptLoaded = false;
+let proxyScriptURL = "proxy/proxy-script.js";
+
+function handleMessage(message, sender) {
+  // only handle messages from the proxy script
+  if (sender.url !=  browser.extension.getURL(proxyScriptURL)) {
+    return;
+  }
+  console.log(message);
+}
+
+function provideCredentialsAsync(details) {
+  return browser.storage.local.get('authCredentials');
+}
+
+// storedSettings includes both proxyServerInfo and authCredentials
+function registerProxyScript(proxyServerInfoAndAuth) {
+  logSettings(proxyServerInfoAndAuth);
+  if (!proxyScriptLoaded) {
+    if (validateProxyServerInfo(proxyServerInfoAndAuth.proxyServerInfo)) {
+      console.log("Registering proxy script");
+
+      // Register the proxy script. Path is relative to manifest.json
+      browser.proxy.register(proxyScriptURL);
+      proxyScriptLoaded = true;
+      sendSettingsToProxyScript(proxyServerInfoAndAuth.proxyServerInfo);
```

Comment:
I would have thought this would race: you might send the message before the proxy script has run and added its listener. But it seems to work...